RawProtocolAuthenticationTest.testAuthenticateWithUsernameAndPassword_with_Invalid_MessageSize

```
java.lang.AssertionError: Expected exception: java.io.EOFException
	at org.junit.internal.runners.statements.ExpectException.evaluate(ExpectException.java:32)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:88)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:82)
	at java.util.concurrent.FutureTask.run(FutureTask.java:277)
	at java.lang.Thread.run(Thread.java:785)
```

https://hazelcast-l337.ci.cloudbees.com/view/Hazelcast%20Maintenance/job/Hazelcast-3.maintenance-IbmJDK1.8/com.hazelcast$hazelcast/75/testReport/junit/com.hazelcast.client.protocol/RawProtocolAuthenticationTest/testAuthenticateWithUsernameAndPassword_with_Invalid_MessageSize/

```
java.lang.AssertionError: Expected exception: java.io.EOFException
	at org.junit.internal.runners.statements.ExpectException.evaluate(ExpectException.java:32)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:88)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:82)
	at java.util.concurrent.FutureTask.run(FutureTask.java:274)
	at java.lang.Thread.run(Thread.java:809)
```

https://hazelcast-l337.ci.cloudbees.com/view/Official%20Builds/job/Hazelcast-3.x-IbmJDK1.7/com.hazelcast$hazelcast/942/testReport/junit/com.hazelcast.client.protocol/RawProtocolAuthenticationTest/testAuthenticateWithUsernameAndPassword_with_Invalid_MessageSize/

```
java.lang.AssertionError: Expected exception: java.io.EOFException
	at org.junit.internal.runners.statements.ExpectException.evaluate(ExpectException.java:32)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:88)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:82)
	at java.util.concurrent.FutureTask.run(FutureTask.java:277)
	at java.lang.Thread.run(Thread.java:785)
```

https://hazelcast-l337.ci.cloudbees.com/view/Hazelcast/job/Hazelcast-3.x-IbmJDK1.8/com.hazelcast$hazelcast/256/testReport/junit/com.hazelcast.client.protocol/RawProtocolAuthenticationTest/testAuthenticateWithUsernameAndPassword_with_Invalid_MessageSize/

--- a/hazelcast/src/test/java/com/hazelcast/client/protocol/RawProtocolAuthenticationTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/client/protocol/RawProtocolAuthenticationTest.java
@@ -20,7 +20,6 @@ import org.junit.Test;
 import org.junit.experimental.categories.Category;
 import org.junit.runner.RunWith;
 
-import java.io.EOFException;
 import java.io.IOException;
 import java.net.InetSocketAddress;
 import java.nio.ByteBuffer;
@@ -28,6 +27,7 @@ import java.nio.channels.SocketChannel;
 import java.util.UUID;
 
 import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
 import static org.junit.Assert.assertNotNull;
 import static org.junit.Assert.assertNull;
 import static org.junit.Assert.assertTrue;
@@ -91,22 +91,7 @@ public class RawProtocolAuthenticationTest {
         final ClientProtocolBuffer byteBuffer = authMessage.buffer();
         channel.write(ByteBuffer.wrap(byteBuffer.byteArray(), 0, authMessage.getFrameLength()));
 
-        final ByteBuffer socketBuffer = ByteBuffer.allocate(4096);
-        ClientMessage clientMessage = ClientMessage.create();
-
-        do {
-            int read = channel.read(socketBuffer);
-            if (read < 0) {
-                throw new EOFException();
-            }
-            socketBuffer.flip();
-            clientMessage.readFrom(socketBuffer);
-            if (socketBuffer.hasRemaining()) {
-                socketBuffer.compact();
-            } else {
-                socketBuffer.clear();
-            }
-        } while (!clientMessage.isComplete());
+        ClientMessage clientMessage = readMessageFromChannel();
 
         assertTrue(clientMessage.isComplete());
 
@@ -138,22 +123,7 @@ public class RawProtocolAuthenticationTest {
         final ClientProtocolBuffer byteBuffer = authMessage.buffer();
         channel.write(ByteBuffer.wrap(byteBuffer.byteArray(), 0, authMessage.getFrameLength()));
 
-        final ByteBuffer socketBuffer = ByteBuffer.allocate(4096);
-        ClientMessage clientMessage = ClientMessage.create();
-
-        do {
-            int read = channel.read(socketBuffer);
-            if (read < 0) {
-                throw new EOFException();
-            }
-            socketBuffer.flip();
-            clientMessage.readFrom(socketBuffer);
-            if (socketBuffer.hasRemaining()) {
-                socketBuffer.compact();
-            } else {
-                socketBuffer.clear();
-            }
-        } while (!clientMessage.isComplete());
+        ClientMessage clientMessage = readMessageFromChannel();
 
         assertTrue(clientMessage.isComplete());
 
@@ -185,22 +155,7 @@ public class RawProtocolAuthenticationTest {
         final ClientProtocolBuffer byteBuffer = authMessage.buffer();
         channel.write(ByteBuffer.wrap(byteBuffer.byteArray(), 0, authMessage.getFrameLength()));
 
-        final ByteBuffer socketBuffer = ByteBuffer.allocate(4096);
-        ClientMessage clientMessage = ClientMessage.create();
-
-        do {
-            int read = channel.read(socketBuffer);
-            if (read < 0) {
-                throw new EOFException();
-            }
-            socketBuffer.flip();
-            clientMessage.readFrom(socketBuffer);
-            if (socketBuffer.hasRemaining()) {
-                socketBuffer.compact();
-            } else {
-                socketBuffer.clear();
-            }
-        } while (!clientMessage.isComplete());
+        ClientMessage clientMessage = readMessageFromChannel();
 
         assertTrue(clientMessage.isComplete());
 
@@ -215,7 +170,7 @@ public class RawProtocolAuthenticationTest {
         assertNull(resultParameters.address);
     }
 
-    @Test(expected = EOFException.class)
+    @Test
     public void testAuthenticateWithUsernameAndPassword_with_Invalid_MessageSize()
             throws IOException, InterruptedException {
 
@@ -229,16 +184,28 @@ public class RawProtocolAuthenticationTest {
                 .encodeRequest(username, pass, null, null, true, ClientTypes.JAVA, (byte) 0);
         authMessage.setCorrelationId(1).addFlag(ClientMessage.BEGIN_AND_END_FLAGS);
 
-        final ClientProtocolBuffer byteBuffer = authMessage.buffer();
-        channel.write(ByteBuffer.wrap(byteBuffer.byteArray()));
+        //set invalid message size
+        authMessage.setFrameLength(ClientMessage.HEADER_SIZE - 1);
 
+        final ClientProtocolBuffer byteBuffer = authMessage.buffer();
+        channel.write(ByteBuffer.wrap(byteBuffer.byteArray(), 0, authMessage.getFrameLength()));
+
+        ClientMessage clientMessage = readMessageFromChannel();
+
+        assertFalse(clientMessage.isComplete());
+    }
+
+    private ClientMessage readMessageFromChannel()
+            throws IOException {
         final ByteBuffer socketBuffer = ByteBuffer.allocate(4096);
         ClientMessage clientMessage = ClientMessage.create();
-
         do {
             int read = channel.read(socketBuffer);
-            if (read < 0) {
-                throw new EOFException();
+            if (read <= 0) {
+                if (read == -1) {
+                    break;
+                }
+                continue;
             }
             socketBuffer.flip();
             clientMessage.readFrom(socketBuffer);
@@ -248,6 +215,6 @@ public class RawProtocolAuthenticationTest {
                 socketBuffer.clear();
             }
         } while (!clientMessage.isComplete());
+        return clientMessage;
     }
-
 }

